hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/io/compress/zlib/BuiltInGzipCompressor.java

```diff
@@ -0,0 +1,256 @@
+
+        state = BuiltInGzipDecompressor.GzipStateLabel.HEADER_BASIC;
+        crc.reset();
+    }
+
+    @Override
+    public boolean finished() {
+        return deflater.finished();
+    }
+
+    @Override
+    public boolean needsInput() {
+        if (state == BuiltInGzipDecompressor.GzipStateLabel.INFLATE_STREAM) {
+            return deflater.needsInput();
+        }
+
+        return (state != BuiltInGzipDecompressor.GzipStateLabel.FINISHED);
+    }
+
+    @Override
+    public int compress(byte[] b, int off, int len) throws IOException {
+        int numAvailBytes = 0;
+
+        // If we are not within uncompression data yet. Output the header.
+        if (state != BuiltInGzipDecompressor.GzipStateLabel.INFLATE_STREAM) {
+            if (userBufLen <= 0) {
+                return numAvailBytes;
+            }
+
+            int outputHeaderSize = writeHeader(b, off, len);
+
+            // Completes header output.
+            if (headerOff == 10) {
+                state = BuiltInGzipDecompressor.GzipStateLabel.INFLATE_STREAM;
+            }
+
+            numAvailBytes += outputHeaderSize;
+
+            if (outputHeaderSize == len) {
+                return numAvailBytes;
+            }
+
+            off += outputHeaderSize;
+            len -= outputHeaderSize;
+        }
+
+        if (state == BuiltInGzipDecompressor.GzipStateLabel.INFLATE_STREAM) {
+            // hand off user data (or what's left of it) to Deflater--but note that
+            // Deflater may not have consumed all of previous bufferload, in which case
+            // userBufLen will be zero
+            if (userBufLen > 0) {
+                deflater.setInput(userBuf, userBufOff, userBufLen);
+
+                crc.update(userBuf, userBufOff, userBufLen);  // CRC-32 is on uncompressed data
+
+                currentInputLen = userBufLen;
+                userBufOff += userBufLen;
+                userBufLen = 0;
+            }
+
+
+            // now compress it into b[]
+            int deflated = deflater.deflate(b, off, len - 8, Deflater.FULL_FLUSH);
+
+            numAvailBytes += deflated;
+            off += deflated;
+            len -= deflated;
+
+            // All current input are processed. Going to output trailer.
+            if (deflater.finished()) {
+                state = BuiltInGzipDecompressor.GzipStateLabel.TRAILER_CRC;
+            } else {
+                return numAvailBytes;
+            }
+        }
+
+        numAvailBytes += writeTrailer(b, off, len);
+
+        return numAvailBytes;
+    }
+
+    @Override
+    public long getBytesRead() {
+        return deflater.getTotalIn();
+    }
+
+    @Override
+    public long getBytesWritten() {
+        return headerBytesWritten + deflater.getTotalOut() + trailerBytesWritten;
+    }
+
+    @Override
+    public void end() { deflater.end(); }
+
+    @Override
+    public void finish() { deflater.finish(); }
+
+    @Override
+    public void reinit(Configuration conf) {
```

Comment:
I think we'll need to reset compression level, strategy etc here too?

Reply:
yea, good catch.

Reply:
this method is sharing a lot of code with the constructor - can we extract them into a util method?